src/BaseVault.sol

```diff
@@ -0,0 +1,662 @@
+        emit Deposit(caller, receiver, assets, shares);
+    }
+
+    /**
+     * @notice Internal function to handle withdrawals.
+     * @param caller The address of the caller.
+     * @param receiver The address of the receiver.
+     * @param owner The address of the owner.
+     * @param assets The amount of assets to withdraw.
+     * @param shares The equivalent amount of shares.
+     */
+    function _withdraw(address caller, address receiver, address owner, uint256 assets, uint256 shares)
+        internal
+        virtual
+    {
+        VaultStorage storage vaultStorage = _getVaultStorage();
+        vaultStorage.totalAssets -= assets;
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        IStrategy(vaultStorage.buffer).withdraw(assets, receiver, address(this));
+
+        _burn(owner, shares);
+        emit Withdraw(caller, receiver, owner, assets, shares);
+    }
+
+    /**
+     * @notice Internal function to convert vault shares to the base asset.
+     * @param asset_ The address of the asset.
+     * @param shares The amount of shares to convert.
+     * @param rounding The rounding direction.
+     * @return (uint256 assets, uint256 baseAssets) The equivalent amount of assets.
+     */
+    function _convertToAssets(address asset_, uint256 shares, Math.Rounding rounding)
+        internal
+        view
+        virtual
+        returns (uint256, uint256)
+    {
+        uint256 assets = shares.mulDiv(totalAssets() + 1, totalSupply() + 10 ** 0, rounding);
+        uint256 baseAssets = _convertBaseToAsset(asset_, assets);
+        return (assets, baseAssets);
+    }
+
+    /**
+     * @notice Internal function to convert assets to shares.
+     * @param asset_ The address of the asset.
+     * @param assets The amount of assets to convert.
+     * @param rounding The rounding direction.
+     * @return (uint256 shares, uint256 baseAssets) The equivalent amount of shares.
+     */
+    function _convertToShares(address asset_, uint256 assets, Math.Rounding rounding)
+        internal
+        view
+        virtual
+        returns (uint256, uint256)
+    {
+        uint256 baseAssets = _convertAssetToBase(asset_, assets);
+        uint256 shares = baseAssets.mulDiv(totalSupply() + 10 ** 0, totalAssets() + 1, rounding);
+        return (shares, baseAssets);
+    }
+
+    /**
+     * @notice Internal function to convert an asset amount to base denomination.
+     * @param asset_ The address of the asset.
+     * @param assets The amount of the asset.
+     * @return uint256 The equivalent amount in base denomination.
+     */
+    function _convertAssetToBase(address asset_, uint256 assets) internal view virtual returns (uint256) {
+        if (asset_ == address(0)) revert ZeroAddress();
+        uint256 rate = IProvider(provider()).getRate(asset_);
+        return assets.mulDiv(rate, 10 ** (_getAssetStorage().assets[asset()].decimals), Math.Rounding.Floor);
```

Comment:
what happens if there's an asset with different decimals from `assets[asset()].decimals`?

same in `_convertBaseToAsset`

Reply:
good catch this is an issue